### PR TITLE
vllm/lora: Patch TPU integration precompilation, dispatch, and circul…

### PR DESCRIPTION
**Pull request by inucool12**
Markdown
**PR Title:**
`[TPU] Resolve TorchXLA Compilation and Import Blockers for Multi-LoRA in v1`

## Purpose
This PR resolves several critical integration blockers preventing Multi-LoRA serving from compiling, initializing, and loading dynamically on TPU / JAX hardware via the TorchXLA (`torchax`) backend in the new vLLM v1 engine.

Specifically, it patches architectural and runtime gaps between the GPU-centric v1 scheduler and JAX/TPU accelerators:
* **Pre-compilation Graph Contexts:** Wraps vLLM's backend pre-compilation and dummy adapter warmups inside conditional `torchax.default_env()` boundaries to allow the TorchXLA compiler to intercept tensor math operations correctly.
* **Strict Dispatcher Assertions:** Bypasses `torchax` type assertions by explicitly casting scalar comparison constants into tensors under the active TPU device context. This prevents the standard PyTorch dispatcher from auto-wrapping constants into raw CPU tensors.
* **Memory Pinning Overrides:** Disables GPU-exclusive memory pinning (`pin_memory=False`) in dynamic host-to-device pipeline copies during compilation setup, which otherwise crashes on TPUs.
* **Circular Dependencies & Resiliency:** Breaks a startup circular import block, and adds robust checks to safely ignore unsupported full-weight outputs instead of crashing the model loading phase.

## Solution Details

**1. Model Runner Precompilation (`vllm/v1/worker/lora_model_runner_mixin.py`)**
* Safely skips dummy LoRA setups during precompilation if `self.lora_manager` resolves to `None` (bypassing early JAX/Flax-nnx crashes).
* Wraps all dummy LoRA mapping additions, activations, and removals in conditionally-imported `torchax.default_env()` contexts to ensure strict compatibility under JAX/XLA graph compilation.

**2. Punica Wrapper Utilities (`vllm/lora/punica_wrapper/utils.py`)**
* Pre-creates scalar fallback values (like `max_loras - 1`) as explicit `torchax` Tensors.
* Dynamically sets `pin_memory=False` in `async_tensor_h2d` to prevent JAX/TPU hardware crashes.

**3. Checkpoint Resiliency (`vllm/lora/lora_model.py`)**
* Wraps key checkpoints parsing in `try-except` blocks to gracefully skip and ignore unsupported full-weight outputs (like `modules_to_save` keys such as `lm_head.weight`) inside PEFT checkpoints during initialization.

**4. Circular Import Fix (`vllm/model_executor/model_loader/bitsandbytes_loader.py`)**
* Relocates the `is_moe_model` import statement from the top module-level to a local import inside `BitsAndBytesModelLoader` methods to break the startup circular import chain.

## Test Plan
We tested the changes end-to-end in the active TPU VM virtual environment (`vllm-venv`) using an instruction-tuned base model `Qwen/Qwen3-4B-Instruct-2507` across 4 TPU tensor-parallel chips, registering three active Hugging Face codec adapters:

```bash
vllm serve Qwen/Qwen3-4B-Instruct-2507 \
  --tensor-parallel-size 4 \
  --dtype bfloat16 \
  --enable-lora \
  --max-loras 3 \
  --max-lora-rank 128 \
  --lora-modules \
      adapter1=voidful/llm-codec-fisher-no-init \
      adapter2=voidful/auv-codec-librispeech \
      adapter3=voidful/unicodec-fisher
```
## Test Results
The v1 server initialized all checkpoints and compiled successfully without JAX model validation crashes, unimplemented wrappers warnings, or XLA memory errors.

The local HTTP server accepted client requests and successfully returned completions payloads, verifying correct backend routing and token counting. (Note: As these are specialized codec/audio adapters, the text completions returned expected codec token strings rather than conversational English. The success metric here is structural routing, zero crashes, and valid token usage generation).

```Bash
curl http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "adapter1",
    "messages": [{"role": "user", "content": "Hello! Who are you?"}],
    "temperature": 0
  }'
```

Example Successful Response (truncated for brevity):
```Bash
JSON
{
  "id": "chatcmpl-9b9d8247e085d144",
  "model": "adapter1",
  "choices": [
    {
      "message": {
        "role": "assistant",
        "content": "Intialized-lndIntialized-lndIntialized... [truncated]"
      },
      "finish_reason": "stop"
    }
  ],
  "usage": {
    "prompt_tokens": 14,
    "total_tokens": 57,
    "completion_tokens": 43
  }
}
```
Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".

- [x] The test plan, such as providing test command.

- [x] The test results, such as pasting the results comparison before and after, or e2e results

- [x] (Optional) The necessary documentation update... *(N/A - backend infrastructure fix, no new models added)*

### CODE REVIEW

File: vllm/lora/punica_wrapper/utils.py [hunk 1]
```diff
@@ -112,24 +112,27 @@ def convert_mapping(
         embedding_indices,
     ]
 
-    indices = async_tensor_h2d(indices_list, dtype=torch.long, device=device)
+    indices = async_tensor_h2d(indices_list, dtype=torch.long, device=device, pin_memory=False)
     prompt_mapping_tensor = async_tensor_h2d(
-        prompt_mapping, dtype=torch.long, device=device
+        prompt_mapping, dtype=torch.long, device=device, pin_memory=False
     )
```
[the remark this answers is not in the document]
Comment (inucool12) [reply vllm/lora/punica_wrapper/utils.py]: Done. Made `pin_memory` conditional on `device.type != "xla"` so that memory pinning is preserved on GPU backends (e.g. CUDA) for optimal DMA page-locked transfers, while avoiding crashes on TPU/XLA.

File: vllm/v1/worker/lora_model_runner_mixin.py [hunk 2]
```diff
@@ -117,7 +117,14 @@ def maybe_setup_dummy_loras(
                 for lora_id in range(1, num_loras + 1)
             }
 
-            with self.lora_manager.dummy_lora_cache():
+            try:
+                import torchax
+                torchax_ctx = torchax.default_env()
+            except ImportError:
+                from contextlib import nullcontext
+                torchax_ctx = nullcontext()
```
[the remark this answers is not in the document]
Comment (inucool12) [reply vllm/v1/worker/lora_model_runner_mixin.py]: Done. Encapsulated the logic in a private `_get_torchax_context(self)` helper that provides a fresh context manager instance (either `torchax.default_env()` or `nullcontext()`) on each entry to avoid sequential reuse issues and remove duplication.

**Comment by inucool12** [comment]
Hello! I am a new contributor. Could a maintainer please take a quick look and add the ready or verified label so the automated tests can run against my draft?
